### DISCUSSION TITLE
Mirroring the sides

### DISCUSSION THREAD

**Discussion by mrrezaie** [ComputationalBiomechanicsLab/opensim-creator#917] [Q&A] [answered]
Hi, I have a model that includes the pelvis and the right lower limb (the femur, tibia, and foot). Now, I plan to create the left lower limb as an exact mirrored version of the right side. By any chance, is there any convenient option in OSC to do this with minimal effort? Something that transforms the geometries, joint centers, coordinates, muscles/ligaments points, and contact geometries automatically. 
Thank you in advance.

**Comment by adamkewley** [comment, marked as answer]
Unfortunately, at time of writing (v0.5.13), there isn't a convenient way to do this that's integrated with OpenSim Creator.

One fundamental issue you might run into (if I'm correct - I'm no expert) is that OpenSim's transform representation can't mirror the frames and mesh data in your model. Allowing something like that would break the transform's right-handed precondition. Therefore, you'll need to redefine frames in the mirrored space, such that the resulting joint coordinates still make sense in the mirror.

> This is one of the reasons why `StationDefinedFrame` was added to OpenSim Core (https://github.com/opensim-org/opensim-core/pull/3694). Using that, frames (rotation + position) can be automatically recalculated by moving stations (positions) in a numerically well-defined way that will work correctly in non-uniform scaling scenarios. I really ought to write a tutorial that describes how to use them in OSC.

Although it's a bit tedious, I think "best" (not good) way to do this right now is manually. I'd do something like:

- Copy the original model so that your edits aren't going to accidentally nuke it
- Open the model in OpenSim Creator. Ensure the `Log` window is open.
- Open the model in a text editor (e.g. Visual Studio) side-by-side
- Copy + paste the bodies you want to mirror below the original bodies in the model (usually, they're all grouped in `bodyset`)
- Rename them as appropriate, OpenSim usually creates freejoints for them
- Do the same for the `JointSet`
- Use a 3rd-party mesh editor (e.g. Blender) to mirror the mesh data (yes, this is unfortunate).

Sorry I can't be of more help right now! My high-level hope is that our model warping feature--whenever that manages to ship--will provide some of the components/algorithms that are required for solving some of these problems (e.g. like providing `StationDefinedFrame`, ways of copying component hierarchies, etc.)
